### ISSUE TITLE
Segmentation fault while training

### ISSUE THREAD

**Issue by WEIIEW97** [HuiZeng/Image-Adaptive-3DLUT#70]
It occured 'Segmentation fault (core dumped)' for cpu version and 'cudaCheckError() failed : invalid device function. Segmentation fault (core dumped)' for CUDA version every time when I trained this network. How could it be solved? Thanks in advance.
![error](https://user-images.githubusercontent.com/47050482/188050130-0f277c25-abbf-44a4-ae76-f996cfa456b1.png)

**Comment by ironheads** [comment]
same problem

**Comment by WEIIEW97** [comment]
> same problem

I fixed this problem by transferring the original CUDA code from c++ to python(numba.cuda). Maybe it avoids some complying bugs. (I tried to use pybind11 but it failed as well)

**Comment by ironheads** [comment]
@WEIIEW97 
hey, sir.
Actually i dive in the cuda c++ file. Then i found that the input  `image` Tensor should be in size `[CHANNEL_SIZE(3), BATCH_SIZE,WIDTH,HEIGHT]`. However, in the python code , the input image is `[BATCH_SIZE, CHANNEL_SIZE(3), WIDTH, HEIGHT]`. So if the **BATCH_SIZE < CHANNEL_SIZE** the c++ cuda code may visit out-of-range memory which may cause this error.

Actually, i was using this code because another work depends on it. and i use it to re-implement that work. So i didn't dive in the python code of this work. But after i change the `compose procedure(compose different LUTs result)` and the `TrilinearInterpolationFunction` of the python code. It seems the  code of that work (which depends on this code) can run normally.

I'd like to know whether you fixed this bug in your python(numba.cuda) code. And it's better if you can share your numbda.cuda file.

Looking forward to your reply.

Thanks.

**Comment by WEIIEW97** [comment]
@ironheads 
hello,

Thanks for your response and sharing! I did not dive deep as you did thus I'd like to appreciate your information.
Attached below is my python(numba.cuda) code. I cannot guarantee the correctness since it is just a basic migration.

Have a nice day.

`python
from __future__ import division
from numba import cuda
import numpy as np
import math
import torch

@cuda.jit(max_registers=40)
def trilinear_forward_kernel(
    nthreads: int,
    lut: float,
    image: float,
    output: float,
    dim: int,
    shift: int,
    binsize: float,
    width: int,
    height: int,
    batch: int,
):
    for index in range(
        cuda.blockIdx.x * cuda.blockDim.x + cuda.threadIdx.x,
        nthreads,
        cuda.blockDim.x * cuda.gridDim.x,
    ):
        r = image[index]
        g = image[index + width * height * batch]
        b = image[index + width * height * batch * 2]

        r_id = math.floor(r / binsize)
        g_id = math.floor(g / binsize)
        b_id = math.floor(b / binsize)

        r_d = math.fmod(r, binsize) / binsize
        g_d = math.fmod(g, binsize) / binsize
        b_d = math.fmod(b, binsize) / binsize

        id000 = int(r_id + g_id * dim + b_id * dim * dim)
        id100 = int(r_id + 1 + g_id * dim + b_id * dim * dim)
        id010 = int(r_id + (g_id + 1) * dim + b_id * dim * dim)
        id110 = int(r_id + 1 + (g_id + 1) * dim + b_id * dim * dim)
        id001 = int(r_id + g_id * dim + (b_id + 1) * dim * dim)
        id101 = int(r_id + 1 + g_id * dim + (b_id + 1) * dim * dim)
        id011 = int(r_id + (g_id + 1) * dim + (b_id + 1) * dim * dim)
        id111 = int(r_id + 1 + (g_id + 1) * dim + (b_id + 1) * dim * dim)

        w000 = (1 - r_d) * (1 - g_d) * (1 - b_d)
        w100 = r_d * (1 - g_d) * (1 - b_d)
        w010 = (1 - r_d) * g_d * (1 - b_d)
        w110 = r_d * g_d * (1 - b_d)
        w001 = (1 - r_d) * (1 - g_d) * b_d
        w101 = r_d * (1 - g_d) * b_d
        w011 = (1 - r_d) * g_d * b_d
        w111 = r_d * g_d * b_d

        output[index] = (
            w000 * lut[id000]
            + w100 * lut[id100]
            + w010 * lut[id010]
            + w110 * lut[id110]
            + w001 * lut[id001]
            + w101 * lut[id101]
            + w011 * lut[id011]
            + w111 * lut[id111]
        )

        output[index + width * height * batch] = (
            w000 * lut[id000 + shift]
            + w100 * lut[id100 + shift]
            + w010 * lut[id010 + shift]
            + w110 * lut[id110 + shift]
            + w001 * lut[id001 + shift]
            + w101 * lut[id101 + shift]
            + w011 * lut[id011 + shift]
            + w111 * lut[id111 + shift]
        )

        output[index + width * height * batch * 2] = (
            w000 * lut[id000 + shift * 2]
            + w100 * lut[id100 + shift * 2]
            + w010 * lut[id010 + shift * 2]
            + w110 * lut[id110 + shift * 2]
            + w001 * lut[id001 + shift * 2]
            + w101 * lut[id101 + shift * 2]
            + w011 * lut[id011 + shift * 2]
            + w111 * lut[id111 + shift * 2]
        )

@cuda.jit(max_registers=40)
def trilinear_backward_kernel(
    nthreads: int,
    image: float,
    image_grad: float,
    lut_grad: float,
    dim: int,
    shift: int,
    binsize: float,
    width: int,
    height: int,
    batch: int,
):
    for index in range(
        cuda.blockIdx.x * cuda.blockDim.x + cuda.threadIdx.x,
        nthreads,
        cuda.blockDim.x * cuda.gridDim.x,
    ):
        r = image[index]
        g = image[index + width * height * batch]
        b = image[index + width * height * batch * 2]

        r_id = math.floor(r / binsize)
        g_id = math.floor(g / binsize)
        b_id = math.floor(b / binsize)

        r_d = math.fmod(r, binsize) / binsize
        g_d = math.fmod(g, binsize) / binsize
        b_d = math.fmod(b, binsize) / binsize

        id000 = int(r_id + g_id * dim + b_id * dim * dim)
        id100 = int(r_id + 1 + g_id * dim + b_id * dim * dim)
        id010 = int(r_id + (g_id + 1) * dim + b_id * dim * dim)
        id110 = int(r_id + 1 + (g_id + 1) * dim + b_id * dim * dim)
        id001 = int(r_id + g_id * dim + (b_id + 1) * dim * dim)
        id101 = int(r_id + 1 + g_id * dim + (b_id + 1) * dim * dim)
        id011 = int(r_id + (g_id + 1) * dim + (b_id + 1) * dim * dim)
        id111 = int(r_id + 1 + (g_id + 1) * dim + (b_id + 1) * dim * dim)

        w000 = (1 - r_d) * (1 - g_d) * (1 - b_d)
        w100 = r_d * (1 - g_d) * (1 - b_d)
        w010 = (1 - r_d) * g_d * (1 - b_d)
        w110 = r_d * g_d * (1 - b_d)
        w001 = (1 - r_d) * (1 - g_d) * b_d
        w101 = r_d * (1 - g_d) * b_d
        w011 = (1 - r_d) * g_d * b_d
        w111 = r_d * g_d * b_d

        cuda.atomic.add(lut_grad, id000, image_grad[index] * w000)
        cuda.atomic.add(lut_grad, id100, image_grad[index] * w100)
        cuda.atomic.add(lut_grad, id010, image_grad[index] * w010)
        cuda.atomic.add(lut_grad, id110, image_grad[index] * w110)
        cuda.atomic.add(lut_grad, id001, image_grad[index] * w001)
        cuda.atomic.add(lut_grad, id101, image_grad[index] * w101)
        cuda.atomic.add(lut_grad, id011, image_grad[index] * w011)
        cuda.atomic.add(lut_grad, id111, image_grad[index] * w111)

        cuda.atomic.add(
            lut_grad, id000 + shift, image_grad[index + width * height * batch] * w000
        )
        cuda.atomic.add(
            lut_grad, id100 + shift, image_grad[index + width * height * batch] * w100
        )
        cuda.atomic.add(
            lut_grad, id010 + shift, image_grad[index + width * height * batch] * w010
        )
        cuda.atomic.add(
            lut_grad, id110 + shift, image_grad[index + width * height * batch] * w110
        )
        cuda.atomic.add(
            lut_grad, id001 + shift, image_grad[index + width * height * batch] * w001
        )
        cuda.atomic.add(
            lut_grad, id101 + shift, image_grad[index + width * height * batch] * w101
        )
        cuda.atomic.add(
            lut_grad, id011 + shift, image_grad[index + width * height * batch] * w011
        )
        cuda.atomic.add(
            lut_grad, id111 + shift, image_grad[index + width * height * batch] * w111
        )

        cuda.atomic.add(
            lut_grad,
            id000 + shift * 2,
            image_grad[index + width * height * batch * 2] * w000,
        )
        cuda.atomic.add(
            lut_grad,
            id100 + shift * 2,
            image_grad[index + width * height * batch * 2] * w100,
        )
        cuda.atomic.add(
            lut_grad,
            id010 + shift * 2,
            image_grad[index + width * height * batch * 2] * w010,
        )
        cuda.atomic.add(
            lut_grad,
            id110 + shift * 2,
            image_grad[index + width * height * batch * 2] * w110,
        )
        cuda.atomic.add(
            lut_grad,
            id001 + shift * 2,
            image_grad[index + width * height * batch * 2] * w001,
        )
        cuda.atomic.add(
            lut_grad,
            id101 + shift * 2,
            image_grad[index + width * height * batch * 2] * w101,
        )
        cuda.atomic.add(
            lut_grad,
            id011 + shift * 2,
            image_grad[index + width * height * batch * 2] * w011,
        )
        cuda.atomic.add(
            lut_grad,
            id111 + shift * 2,
            image_grad[index + width * height * batch * 2] * w111,
        )


def trilinear_forward(
    lut: float,
    image: float,
    output: float,
    lut_dim: int,
    shift: int,
    binsize: float,
    width: int,
    height: int,
    batch: int,
):
    k_thread_per_block = 1024
    output_size = height * width * batch

    trilinear_forward_kernel[
        int((output_size + k_thread_per_block - 1) / k_thread_per_block), k_thread_per_block
    ](output_size, lut, image, output, lut_dim, shift, binsize, width, height, batch)


def trilinear_backward(
    image: float,
    image_grad: float,
    lut_grad: float,
    lut_dim: int,
    shift: int,
    binsize: float,
    width: int,
    height: int,
    batch: int,
):
    k_thread_per_block = 1024
    output_size = height * width * batch

    trilinear_backward_kernel[
        int((output_size + k_thread_per_block - 1) / k_thread_per_block), k_thread_per_block
    ](
        output_size,
        image,
        image_grad,
        lut_grad,
        lut_dim,
        shift,
        binsize,
        width,
        height,
        batch,
    )


def trilinear_forward_cuda(
    lut: torch.Tensor,
    image: torch.Tensor,
    output: torch.Tensor,
    lut_dim: int,
    shift: int,
    binsize: float,
    width: int,
    height: int,
    batch: int,
):
    lut_flat = lut.flatten().float().detach()
    image_flat = image.flatten().float().detach()
    output_flat = output.flatten().float().detach()

    trilinear_forward(
        lut_flat, image_flat, output_flat, lut_dim, shift, binsize, width, height, batch
    )


def trilinear_backward_cuda(
    image: torch.Tensor,
    image_grad: torch.Tensor,
    lut_grad: torch.Tensor,
    lut_dim: int,
    shift: int,
    binsize: float,
    width: int,
    height: int,
    batch: int,
):
    image_grad_flat = image_grad.flatten().float().detach()
    image_flat = image.flatten().float().detach()
    lut_grad_flat = lut_grad.flatten().float().detach()

    trilinear_backward(
        image_flat,
        image_grad_flat,
        lut_grad_flat,
        lut_dim,
        shift,
        binsize,
        width,
        height,
        batch,
    )

`

**Comment by WEIIEW97** [comment]
@ironheads
hello,

Thanks for your response and sharing! I did not dive deep as you did thus I'd like to appreciate your information.
Attached below is my python(numba.cuda) code. I cannot guarantee robustness since it is just a basic migration. 

[trilinearcuda.zip](https://github.com/HuiZeng/Image-Adaptive-3DLUT/files/9580378/trilinearcuda.zip)

**Comment by ironheads** [comment]
@WEIIEW97 
Thanks for your sharing.
If It is a basic migration. I think there are some codes need to modify in the original training/evaluation python code.
Some Modifications shows as following.
```python
# LUT0 = Generator3DLUT_identity()
# LUT1 = Generator3DLUT_zero()
# LUT2 = Generator3DLUT_zero()
#...
#...
# img = some images in  the dataset whose shape is [batch_size,3,width,height], make sure that the values are in range [0,1] (My segmentation fault comes from this)
# the following code comes from image_adaptive_lut_train_paired.py
pred = classifier(img).squeeze() # the img is still [batch_size,3,width,height]
# then you should modify the codes as following
new_img = img.permute(1,0,2,3).contiguous()
gen_A0 = LUT0(new_img)
gen_A1 = LUT1(new_img)
gen_A2 = LUT2(new_img)
combine_A = new_img.new(new_img.size())
    for b in range(img.size(1)):
        combine_A[:,b,:,:] = pred[b,0] * gen_A0[:,b,:,:] + pred[b,1] * gen_A1[:,b,:,:] + pred[b,2] * gen_A2[:,b,:,:] #+ pred[b,3] * gen_A3[:,b,:,:] + pred[b,4] * gen_A4[:,b,:,:]

result_A = combine_A.permute(1,0,2,3) #get the [batch_size,3,width,height] combined image
# the key is make the LUT's input in shape [3,batch_size,width,width], there maybe some other codes need to be modified when using LUT. I don't list them all because i do not use all the python codes in this work.

# another important modification is TrilinearInterpolationFunction, it should be modified as following.
class TrilinearInterpolationFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, lut, x):
        x = x.contiguous()
        output = x.new(x.size())
        dim = lut.size()[-1]
        shift = dim ** 3
        binsize = 1.000001 / (dim-1)
        W = x.size(2)
        H = x.size(3)
        batch = x.size(1)
        # print(batch)
        assert 1 == trilinear.forward(lut, 
                                      x, 
                                      output,
                                      dim, 
                                      shift, 
                                      binsize, 
                                      W, 
                                      H, 
                                      batch)

        int_package = torch.IntTensor([dim, shift, W, H, batch])
        float_package = torch.FloatTensor([binsize])
        variables = [lut, x, int_package, float_package]
        
        ctx.save_for_backward(*variables)
        
        return lut, output
    
    @staticmethod
    def backward(ctx, lut_grad, x_grad):
        
        lut, x, int_package, float_package = ctx.saved_variables
        dim, shift, W, H, batch = int_package
        dim, shift, W, H, batch = int(dim), int(shift), int(W), int(H), int(batch)
        binsize = float(float_package[0])
            
        assert 1 == trilinear.backward(x, 
                                       x_grad, 
                                       lut_grad,
                                       dim, 
                                       shift, 
                                       binsize, 
                                       W, 
                                       H, 
                                       batch)
        return lut_grad, x_grad
```

all the modification aims to make the input of LUT in shape [3,batch_size,width,height] and then reshape it into [batch_size,3,width,height].

Another important thing is that the input values should be in range [0,1]

I don't know whether this will help If your code fails when batch_size > 1.
but if you set batch_size > 1, you should modify the code. because when batch_size > 1, the code is not a implementation of LUT.

**Comment by WEIIEW97** [comment]
@ironheads 
可以的, 感谢老哥! 我去瞅一眼
